sway/desktop/launcher.c

```diff
@@ -79,45 +82,83 @@ struct sway_workspace *root_workspace_for_pid(pid_t pid) {
 			if (pid == _pw->pid) {
 				pw = _pw;
 				sway_log(SWAY_DEBUG,
-						"found pid_workspace for pid %d, workspace %s",
-						pid, pw->workspace);
-				goto found;
+					"found %s match for pid %d: %s",
+					node_type_to_str(pw->node->type), pid, node_get_name(pw->node));
+				break;
 			}
 		}
 		pid = get_parent_pid(pid);
 	} while (pid > 1);
 
-found:
-	if (pw && pw->workspace) {
-		ws = workspace_by_name(pw->workspace);
-
-		if (!ws) {
-			sway_log(SWAY_DEBUG,
-					"Creating workspace %s for pid %d because it disappeared",
-					pw->workspace, pid);
-
-			struct sway_output *output = pw->output;
-			if (pw->output && !pw->output->enabled) {
+	if (pw) {
+		switch (pw->node->type) {
+		case N_CONTAINER:
+			// Unimplemented
+			// TODO: add container matching?
+			ws = pw->node->sway_container->pending.workspace;
+			break;
+		case N_WORKSPACE:
+			ws = pw->node->sway_workspace;
+			break;
+		case N_OUTPUT:
+			output = pw->node->sway_output;
+			ws = workspace_by_name(pw->name);
+			if (!ws) {
 				sway_log(SWAY_DEBUG,
-						"Workspace output %s is disabled, trying another one",
-						pw->output->wlr_output->name);
-				output = NULL;
+						"Creating workspace %s for pid %d because it disappeared",
+						pw->name, pid);
+				if (!output->enabled) {
+					sway_log(SWAY_DEBUG,
+							"Workspace output %s is disabled, trying another one",
+							output->wlr_output->name);
+					output = NULL;
+				}
+				ws = workspace_create(output, pw->name);
 			}
-
-			ws = workspace_create(output, pw->workspace);
+			break;
+		case N_ROOT:
+			ws = workspace_create(NULL, pw->name);
+			break;
 		}
-
 		pid_workspace_destroy(pw);
 	}
 
 	return ws;
 }
 
-static void pw_handle_output_destroy(struct wl_listener *listener, void *data) {
-	struct pid_workspace *pw = wl_container_of(listener, pw, output_destroy);
-	pw->output = NULL;
-	wl_list_remove(&pw->output_destroy.link);
-	wl_list_init(&pw->output_destroy.link);
+static void pw_handle_node_destroy(struct wl_listener *listener, void *data) {
+	struct pid_workspace *pw = wl_container_of(listener, pw, node_destroy);
+	switch (pw->node->type) {
+	case N_CONTAINER:
+		// Unimplemented
+		break;
+	case N_WORKSPACE:;
+		struct sway_workspace *ws = pw->node->sway_workspace;
+		wl_list_remove(&pw->node_destroy.link);
+		wl_list_init(&pw->node_destroy.link);
+		// We want to save this ws name to recreate later, hopefully on the
+		// same output
+		pw->name = strdup(ws->name);
```

Comment:
We've already done this in `root_record_workspace_pid()` I think?

Reply:
Hm this is the same error as in the last review, but I didn't go far enough back when applying the fix. What's missing is a free-ing the old name, we still want the strdup I think.

What I'm doing here is updating pw->name in case it has changed since the process was launched. In order to preserve sway's existing behavior, I'm trying to accommodate the following scenario that currently works in sway:

1. User has workspace "A"
2. User launches process 123
3. User renames workspace "A" -> "B"
4. User switches to another workspace and an empty workspace "B" is reaped
5. Process 123 maps a window, it should be pid matched to workspace "B"
6. Workspace "B" is created

Previously all pid matches were handled by recording the name and looking it up when the window was mapped. When a workspace was renamed a helper function walked the list of pid matches and updated the tracked workspace names too. Now in b2133d968 (launcher: track workspaces by node) I dropped that and matches are handled directly by node, so it doesn't matter if the workspace name has changed ... unless it was changed before it was destoyed and we need to recreate a workspace with the updated name. So I'm recording it here for that purpose — this variable is now "name of the workspace to create in case the one we intended to match has been destroyed" .